### PR TITLE
Kyori Components

### PR DESCRIPTION
**Pull request by rexlManu**
My pr removes mostly of the legacy handling of prior version 1.17 and introduces kyori components instead of strings.

I don't think it makes sense, merging  it with the current main branch. My goal is to provide a "newer" version with kyori components in a separate branch or even repo.

Anyway I created this pr, so other folks can just copy the classes with kyori as alternative instead of strings.

Have a good day!

### CODE REVIEW

**Comment by MrMicky-FR** [comment]
Hi, thanks for your pull request, please just keep the current formatting (4 spaces) because currently the diff is impossible to review

**Comment by rexlManu** [comment]
Oh yea, sorry for that. I'm used to just apply the google java styling conventions. Fixed it.

**Comment by Doc94** [comment]
Hi, any update about this feature?

**Comment by MrMicky-FR** [comment]
> Hi, any update about this feature?

Unfortunately not yet: Adventure support is still planned but I have no ETA, and this PR is almost a complete rewrite of FastBoard (drops support for lots of old versions and remove Bukkit/Spigot support) so I cannot directly merge it without introducing breaking changes for almost all FastBoard users 😐

**Comment by Doc94** [comment]
> > Hi, any update about this feature?
> 
> Unfortunately not yet: Adventure support is still planned but I have no ETA, and this PR is almost a complete rewrite of FastBoard (drops support for lots of old versions and remove Bukkit/Spigot support) so I cannot directly merge it without introducing breaking changes for almost all FastBoard users 😐

okay, and understand maybe if this PR support the "legacy" method can be great.
unless exists a form to transform the component to the legacy text for use with this lib?

**Comment by MrMicky-FR** [comment]
> > > Hi, any update about this feature?
> > 
> > 
> > Unfortunately not yet: Adventure support is still planned but I have no ETA, and this PR is almost a complete rewrite of FastBoard (drops support for lots of old versions and remove Bukkit/Spigot support) so I cannot directly merge it without introducing breaking changes for almost all FastBoard users 😐
> 
> okay, and understand maybe if this PR support the "legacy" method can be great. unless exists a form to transform the component to the legacy text for use with this lib?

You can use the [Legacy Serializer](https://docs.adventure.kyori.net/serializer/legacy.html) to convert components from/to strings

**Comment by rexlManu** [comment]
Hey, it depends on the use case but I think it doesn't make sense to support legacy versions because adventure is non existence prior >1.16. So thats why I dropped it and yea it doesnt' make sense to merge it, I just open this pr to provide a alternative version.

**Comment by Doc94** [comment]
> > > > Hi, any update about this feature?
> > > 
> > > 
> > > Unfortunately not yet: Adventure support is still planned but I have no ETA, and this PR is almost a complete rewrite of FastBoard (drops support for lots of old versions and remove Bukkit/Spigot support) so I cannot directly merge it without introducing breaking changes for almost all FastBoard users 😐
> > 
> > 
> > okay, and understand maybe if this PR support the "legacy" method can be great. unless exists a form to transform the component to the legacy text for use with this lib?
> 
> You can use the [Legacy Serializer](https://docs.adventure.kyori.net/serializer/legacy.html) to convert components from/to strings

Nice, this works perfect for convert.



> Hey, it depends on the use case but I think it doesn't make sense to support legacy versions because adventure is non existence prior >1.16. So thats why I dropped it and yea it doesnt' make sense to merge it, I just open this pr to provide a alternative version.

yeah not sure what can be better, if add support to the two things and make maybe component convert to legacy text for allow old server can use adventure in old systems (?, or make the breaking and drop backcompatibility (?

**Comment by rexlManu** [comment]
The problem is that old versions don't have adventure. So it need to be 2 versions of the lib anyway.

**Review by portlek — CHANGES_REQUESTED** [review]

File: src/main/java/fr/mrmicky/fastboard/FastBoard.java [hunk 1]
```diff
@@ -509,63 +488,32 @@ private void sendTeamPacket(int score, TeamMode mode) throws Throwable {
             throw new UnsupportedOperationException();
         }
 
-        int maxLength = hasLinesMaxLength() ? 16 : 1024;
         Object packet = PACKET_SB_TEAM.invoke();
 
         setField(packet, String.class, this.id + ':' + score); // Team name
-        setField(packet, int.class, mode.ordinal(), VERSION_TYPE == VersionType.V1_8 ? 1 : 0); // Update mode
+        setField(packet, int.class, mode.ordinal(), 1); // Update mode
```
[the remark this answers is not in the document]
Comment (rexlManu) [reply src/main/java/fr/mrmicky/fastboard/FastBoard.java]: Thanks for the report. Fixed it.

**Comment by rexlManu** [comment]
I was thinking again about that. We could support both when we use object class and check on building if it's components and if the server is running a new version otherwise convert them to legacy text. This change would introduce a lot of reflection and a possible dependency with kyori if components are used.

I could create that and PR it, but just want to know if that's even needed.

**Comment by portlek** [comment]
you can make it work with 1 abstract class and 2 implementation both string and component for the packet writing parts i guess.

**Comment by rexlManu** [comment]
> you can make it work with 1 abstract class and 2 implementation both string and component for the packet writing parts i guess.

I would do it like that if it's fine when api break changes are okay.

**Comment by MrMicky-FR** [comment]
> > you can make it work with 1 abstract class and 2 implementation both string and component for the packet writing parts i guess.
> 
> I would do it like that if it's fine when api break changes are okay.

Hi, I already tried splitting FastBoard into an interface with two implementations (String and Component) but I couldn’t get good results without lot of duplicated code, which would highly impact maintainability, but I will give it a new try this week!

About a PR for this split, yes it would be great, but the key considerations would be to not have too much duplicate code and keep the current features, like legacy versions compatibility, almost single-class import (including without Maven/Gradle, even if importing through it is recommended), Bukkit (non-Paper) support, … 
Breaking changes are fine, as it would be released as a v2.0.0.

**Comment by MrMicky-FR** [comment]
> you can make it work with 1 abstract class and 2 implementation both string and component for the packet writing parts i guess.

I managed to get a pretty good result, I've pushed it on the [dev branch](https://github.com/MrMicky-FR/FastBoard/tree/dev), not tested yet but at least it compiles. Feel free to add feedback in the commit, will open a draft PR when tested :)

**Comment by rexlManu** [comment]
Looks good, I would suggust to add a check if the server is running on paper otherwise convert it  back to strings with the legacy serializer from kyori.

That would fix the problem when a author wants to use kyori but the plugin needs to run on like 1.12, the components will be serialized back to old color codes, if we want to keep the support from 1.8 until 1.19.4. Otherwise its fine to drop support for that for kyori components.

**Comment by rexlManu** [comment]
Due the implementation and the currently draft pr from Micky, I will close this pr.

**Pull request closed without merging**